src/main/resources/META-INF/rewrite/no-guava.yml

```diff
@@ -64,19 +65,53 @@ name: org.openrewrite.java.migrate.guava.PreferJavaUtilFunction
 displayName: Prefer `java.util.function.Function`
 description: Prefer `java.util.function.Function` instead of using `com.google.common.base.Function`.
 tags:
-  - guava, RSPEC-4738
+  - guava
+  - RSPEC-4738
 recipeList:
   - org.openrewrite.java.ChangeType:
       oldFullyQualifiedTypeName: com.google.common.base.Function
       newFullyQualifiedTypeName: java.util.function.Function
 
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: org.openrewrite.java.migrate.guava.PreferJavaUtilOptional
+displayName: Prefer `java.util.Optional`
+description: Prefer `java.util.Optional` instead of using `com.google.common.base.Optional`.
+tags:
+  - guava
+  - RSPEC-4738
+recipeList:
+  - org.openrewrite.java.migrate.guava.PreferJavaUtilOptionalOrSupplier
+  - org.openrewrite.java.migrate.guava.PreferJavaUtilOptionalOrElseNull
+  - org.openrewrite.java.migrate.guava.NoGuavaOptionalFromJavaUtil
+  - org.openrewrite.java.migrate.guava.NoGuavaOptionalToJavaUtil
+  - org.openrewrite.java.ChangeMethodName:
+      methodPattern: com.google.common.base.Optional absent()
+      newMethodName: empty
+  - org.openrewrite.java.ChangeMethodName:
+      methodPattern: com.google.common.base.Optional fromNullable(..)
+      newMethodName: ofNullable
+  - org.openrewrite.java.ChangeMethodName:
+      methodPattern: com.google.common.base.Optional or(com.google.common.base.Supplier)
+      newMethodName: orElseGet
```

Comment:
This recipe works nicely if the supplier is specified as a lambda but fails when the supplier is represented by a variable / parameter of type `com.google.common.base.Supplier`. I think we can keep this as a known limitation for now (we could just add a test case for it so we don't forget).

Reply:
I just noticed that we also have recipes to migrate from Guava Supplier and Function to the corresponding JDK types. Should we add a test for this?

Reply:
I just tested this myself (by replacing the recipe used by `PreferJavaUtilOptionalTest` with `org.openrewrite.java.migrate.guava.NoGuava`) and it indeed works 😃  But the `transformToMap()` test failed. This test however works if we in `no-guava.yml` move the `PreferJavaUtilFunction` recipe below the `PreferJavaUtilOptional` recipe.

Reply:
Thanks! Hadn't thought about those interacting indeed; now tested and fixed in 06a5a8d.

As for testing Supplier and Function: those are implemented as calls to ChangeType only, so I don't think we have to test those extensively here. The fix would likely be there for these functional interfaces.
https://github.com/openrewrite/rewrite-migrate-java/blob/9a74c66bf6698dd437dedb3fcd0f74f07fe9bc03/src/main/resources/META-INF/rewrite/no-guava.yml#L85-L95

Reply:
Also added a test case `orSupplierToOrElseGetWithSupplierArgument` in bb48f4c to match the test case you mentioned (I think).